### PR TITLE
Const-to-pattern-to-MIR cleanup

### PR DESCRIPTION
**Pull request by RalfJung**
Now that all uses of constants without structural equality are hard errors, there's a bunch of cleanup we can do in the code that handles patterns: we can always funnel patterns through valtrees first (rather than having a fallback path for when valtree construction fails), and we can make sure that if we emit a `PartialEq` call it is not calling anything user-defined.

To keep the error messages the same, I made valtree construction failures return the information of *which* type it is that cannot be valtree'd. `search_for_structural_match_violation` is now not needed any more at all, so I removed it.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_mir_build/src/thir/pattern/mod.rs [hunk 1]
```diff
@@ -575,63 +574,39 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
...
-            Ok(const_) => {
-                let pattern = self.const_to_pat(const_, id, span);
-
-                if !is_associated_const {
-                    return pattern;
-                }
+        let c = ty::Const::new_unevaluated(
+            self.tcx,
+            ty::UnevaluatedConst { def: instance.def_id(), args: instance.args },
+        );
```
Comment (RalfJung) [inline compiler/rustc_mir_build/src/thir/pattern/mod.rs:580]: This code looks a bit odd... we create an instance only to immediately take it apart again. Can this be simplified further? We have `def_id` and `args` already before the `let instance = ...`, can we just use those directly? But there are some specific error cases up there that I don't know how to handle then, in particular `AssocConstInPattern`.

Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/mod.rs:580]: If I remove the `Instance::try_resolve`, then the only change in the tests is that "associated consts cannot be referenced in patterns" becomes "constant pattern depends on a generic parameter". Which seems okay? It *does* depend on a generic parameter:
```rust
pub fn test<A: Foo, B: Foo>(arg: EFoo, A::X: EFoo) {
    //~^ ERROR associated consts cannot be referenced in patterns
    let A::X = arg;
    //~^ ERROR associated consts cannot be referenced in patterns
}
```
If the associated const can be resolved, I am fairly sure we already accept that code, so the general claim "associated consts cannot be referenced in patterns" is already not true I think.

Any objections to changing the error message for that case?

File: compiler/rustc_mir_build/src/thir/pattern/mod.rs [hunk 2]
```diff
@@ -548,37 +548,8 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
             _ => return pat_from_kind(self.lower_variant_or_leaf(res, id, span, ty, vec![])),
         };
 
-        // Use `Reveal::All` here because patterns are always monomorphic even if their function
-        // isn't.
-        let param_env_reveal_all = self.param_env.with_reveal_all_normalized(self.tcx);
-        // N.B. There is no guarantee that args collected in typeck results are fully normalized,
-        // so they need to be normalized in order to pass to `Instance::resolve`, which will ICE
-        // if given unnormalized types.
```
Comment (RalfJung) [inline compiler/rustc_mir_build/src/thir/pattern/mod.rs:557]: I was not able to reproduce any ICEs here with the new version that directly construct an unevaluated consts from the args. So I removed the normalization step. If that can cause ICEs I am sure matthiaskrgr's fuzzer will find them quickly, and then we'll have a test. ;)

File: compiler/rustc_middle/src/thir.rs [hunk 3]
```diff
@@ -783,16 +783,13 @@ pub enum PatKind<'tcx> {
     },
 
     /// One of the following:
-    /// * `&str` (represented as a valtree), which will be handled as a string pattern and thus
+    /// * `&str`/`&[u8]` (represented as a valtree), which will be handled as a string pattern and thus
```
Comment (fmease) [inline compiler/rustc_middle/src/thir.rs:786]:
```suggestion
    /// * `&str`/`&[u8]` (represented as a valtree), which will be handled as a string/slice pattern and thus
```

I assume (I'm not particularly well-versed in `rustc_pattern_analysis`)

File: compiler/rustc_middle/src/query/erase.rs [hunk 4]
```diff
@@ -157,9 +157,10 @@ impl EraseType for Result<mir::ConstValue<'_>, mir::interpret::ErrorHandled> {
     type Result = [u8; size_of::<Result<mir::ConstValue<'static>, mir::interpret::ErrorHandled>>()];
 }
 
-impl EraseType for Result<Option<ty::ValTree<'_>>, mir::interpret::ErrorHandled> {
-    type Result =
-        [u8; size_of::<Result<Option<ty::ValTree<'static>>, mir::interpret::ErrorHandled>>()];
+impl EraseType for Result<Result<ty::ValTree<'_>, Ty<'_>>, mir::interpret::ErrorHandled> {
+    type Result = [u8; size_of::<
+        Result<Result<ty::ValTree<'static>, Ty<'static>>, mir::interpret::ErrorHandled>,
```
Comment (lcnr) [inline compiler/rustc_middle/src/query/erase.rs:162]: use the `mir::interpret::EvalToValTreeResult` type alias here :thinking:

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs [hunk 5]
```diff
@@ -91,116 +89,55 @@ impl<'tcx> ConstToPat<'tcx> {
...
-                // The only way valtree construction can fail without the structural match
-                // checker finding a violation is if there is a pointer somewhere.
-                let e = self.tcx().dcx().emit_err(PointerPattern { span: self.span });
-                let kind = PatKind::Error(e);
-                return Box::new(Pat { span: self.span, ty: cv.ty(), kind });
-            }
+        // Convert the valtree to a const.
+        let inlined_const_as_pat = self.valtree_to_pat(valtree, ty);
 
+        if self.saw_const_match_error.get().is_none() {
```
Comment (lcnr) [inline compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs:134]: can this check `inlined_const_as_pat.references_error()` instead? I think we only set `saw_const_match_error` when also replacing that pattern with `Pat::Error`

**Comment by lcnr** [comment]
really happy with these changes :blush: I haven't interacted with this area recently so I don't feel fully confident in signing this off myself, would like oli to do a final review

File: compiler/rustc_mir_build/src/thir/pattern/mod.rs (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_mir_build/src/thir/pattern/mod.rs]: I think that change is fine :+1:

File: compiler/rustc_middle/src/query/erase.rs (hunk 4 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/query/erase.rs:162]: Is that a good idea? This seems like a very dangerous trait to implement, it just transmutes some lifetimes around, so maybe it'd be dangerous to just change the type alias without realizing that this also changes this here.

Unfortunately the `EraseType` trait has no comment at all, and to my great surprise it is a safe trait, so I really don't understand what happens here.

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs (hunk 5 above)
Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs:134]: I had no idea that function existed.^^ Yes, that's a great way to get rid of this annoying `Cell`.

Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs]: Hm, it doesn't seem to work though. The error is shown now even if we already previously showed other errors, e.g.
```
error: to use a constant of type `MyType` in a pattern, `MyType` must be annotated with `#[derive(PartialEq)]`
  --> $DIR/const-partial_eq-fallback-ice.rs:14:12
   |
LL |     if let CONSTANT = &&MyType {
   |            ^^^^^^^^
   |
   = note: the traits must be derived, manual `impl`s are not sufficient
   = note: see https://doc.rust-lang.org/stable/std/marker/trait.StructuralPartialEq.html for details

error: to use a constant of type `&&MyType` in a pattern, the type must implement `PartialEq`
  --> $DIR/const-partial_eq-fallback-ice.rs:14:12
   |
LL |     if let CONSTANT = &&MyType {
   |            ^^^^^^^^

error: aborting due to 2 previous errors
```

Comment (lcnr) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs]: that is very confusing to me and I can't see how that would happen :thinking: all the cases which previously set `saw_const_match_error` returns as `PatKind::Error` and afaict we never discard subpattern :thinking: could you revert the commit and test with `assert_eq!(!inlined_const_as_pat.references_error(), saw_const_match_error.get().is_none(), "pat={inlined_const_as_pat:?}");`

Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs]: The problem is that `TypeVisitable` does not have a dedicated method for `ErrorGuaranteed`, so only errors in types, constants and lifetimes are actually found.

Comment (lcnr) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs]: ah :thinking: should we add that method? :thinking: it feels useful :blush:

Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs]: https://github.com/rust-lang/rust/pull/127808 should fix this

**Review by oli-obk — APPROVED** [review]
r=me,lcnr after https://github.com/rust-lang/rust/pull/127808

**Pull request merged**